### PR TITLE
add datastream support

### LINKED ISSUE #766: [FEATURE] Support dataStream inside OpensearchIndexTemplate
To allow [data streams](https://opensearch.org/docs/latest/im-plugin/data-streams/) to be created directly from an index template, it is necessary to support this in the `OpensearchIndexTemplate` resource.

Example OpensearchIndexTemplate:
```yaml
apiVersion: opensearch.opster.io/v1
kind: OpensearchIndexTemplate
metadata:
  name: logs
spec:
  opensearchCluster:
    name: test
  name: logs
  indexPatterns:
    - "logs-*"
  dataStream: {}
```

### PR DESCRIPTION
**Pull request by cgroschupp**
### Description
Add DataStream support to IndexTemplate.

Close  #766

### CODE REVIEW

**Comment by cthtrifork** [comment]
Great! I would love this feature in the next coming release! (perhaps 2.5.2? https://github.com/opensearch-project/opensearch-k8s-operator/issues/749)

**Comment by cgroschupp** [comment]
@cthtrifork As this is a new feature, I think the 2.3.0 version is better.

@idanl21 Can you please review this pull request?

**Comment by prudhvigodithi** [comment]
Thanks @cgroschupp for your contribution, before we get this PR merged, can you please open an issue summarizing the datastream support in detail? This way your contribution is tracked in the issue and later we can link this PR to the issue.

Adding @bbarani @salyh @jochenkressin @pchmielnik @swoehrl-mw

**Comment by cgroschupp** [comment]
@prudhvigodithi done

**Comment by prudhvigodithi** [comment]
Hey @cgroschupp we can push this to `v2.6.0` https://github.com/opensearch-project/opensearch-k8s-operator/issues/776, can you please add unit tests to this change ?
Adding @swoehrl-mw @salyh @cthtrifork can you please add your review?
Thanks

**Comment by cgroschupp** [comment]
@prudhvigodithi  done

File: opensearch-operator/api/v1/opensearch_index_types.go [hunk 1]
```diff
@@ -4,6 +4,16 @@ import (
...
 )
 
+type OpensearchDatastreamTimestampFieldSpec struct {
+	// Name of the field that are used for the DataStream
+	Name string `json:"name"`
+}
+
+type OpensearchDatastreamSpec struct {
+	// TimestampField for dataStream
+	TimestampField OpensearchDatastreamTimestampFieldSpec `json:"timestamp_field,omitempty"`
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_index_types.go:14]: Every other field is camelCase, so this should as well to be consistent

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
One gripe about naming, everything else looks good

File: opensearch-operator/api/v1/opensearch_index_types.go [hunk 2]
```diff
@@ -4,6 +4,16 @@ import (
...
 )
 
+type OpensearchDataStreamTimestampFieldSpec struct {
+	// Name of the field that are used for the DataStream
+	Name string `json:"name"`
+}
+
+type OpensearchDataStreamSpec struct {
+	// TimestampField for dataStream
+	TimestampField OpensearchDataStreamTimestampFieldSpec `json:"timestamp_field,omitempty"`
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_index_types.go:14]: Sorry for not being clearer, my camelCase comment was primarily aimed at `timestamp_field` (should be `timestampField`). Even though in the opensearch api it is snake_case, the operator CRDs use camelCase all around

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]

**Review by swoehrl-mw — APPROVED** [review]

**Pull request merged**